TEE.md

```diff
@@ -0,0 +1,79 @@
+
+For the MPC network, the security guarantees provided by TEEs are attractive mainly because they allow the use of more relaxed threat models, unlocking performant MPC protocols.
+
+TEEs provide their security guarantees by restricting how anything outside of the TEE can interact with the code running inside the TEE. This is great to protect against malicious actors, but it also restricts the honest actors. It has to be expected that debugging and handling of emergencies will become much more difficult compared to running an MPC node outside of a TEE.
+
+This is a good moment to remind the reader that threshold crypotgrapy requires `threshold` operational nodes. - anything less than that, and the protocol finds itself incapacitated, not able to move any funds.
+
+Therefore, Near one will roll-out their TEE implementation in two phases:
+- Soft Launch: node operators are invited to try out and run their node inside a TEE. This will happen in a coordinated manner.
+- Hard Launch: node operators will be required to run their node inside a TEE.
+
+In order to protect the network from the worst case scenario (complete loss of funds due to loss of key shares or an incapacitated network), a disaster recovery plan is prepared.
+
+
+## Disaster Recovery
+Distaster recovery is a plan intended to prevent a permanent loss of the signing quorum.
+As long as the secret shares of the MPC nodes are securely backed-up outside of the TEE environment in which the node is running, it highly likely that the network will be able to recover from otherwise catastrophic events.
+
+Therefore, the disaster recovery plan aims at establishing a secure mechanism allowing to back-up the secret shares held by the nodes. It encompasses two steps:
+1. (Back-up Generation): export the private keys from the TEE in a secure manner.
+2. (Recovery): when required, securely import the private keys into the node.
+
+### Back-up mechanism:
+On a high-level, the back-up mechanism works like this:
+
+1. The node running inside the TEE and the back-up recipient must agree on a key used to encrypt the secrets.
+2. The node encrypts the sensitive secret shares with the derived key.
+3. The node exports the encrypted secret shares.
+
+#### Agreeing on an encryption key
+A pair-wise key establishement scheme can be leveraged to establish a symmetric key. A suitable implementation can be found in [libsodium](https://doc.libsodium.org/) and its rust wrapper [libsodium-rs](https://docs.rs/crate/libsodium-rs/latest). Specifically, the **key exchange protocol based on X25519** would be a good fit for the given set-up (c.f. [libsodium](https://doc.libsodium.org/key_exchange) and [libsodium-rs](https://docs.rs/libsodium-rs/latest/libsodium_rs/crypto_kx/index.html)).
+
+_Note: This key exchange protocol is on Curve25519, which is different from the Edwards curve used by NEAR._
```

Comment:
Why do we use a different curve? Any Diffie-Hellman based key exchange should work for most curves anyway.

That said, Curve25519 is a nice curve and I'm happy to use it - but if we already have keys on another curve it makes more sense to reuse them.

Reply:
Well, you know more than I. I thought Curve25519 was the twisted Edwards curve used in the Ed25519 scheme - but TIL they are different curves.